File: tuan2/PhanC/bai48.py
import sys
import math

def kc_euclid(u, v):
    return math.sqrt(sum((u[k] - v[k]) ** 2 for k in range(len(u))))

def kc_manhattan(u, v):
    return sum(abs(u[k] - v[k]) for k in range(len(u)))

def phan_loai(u, D, kc_func, skip_id=-1):
    kc_min = float("inf")
    du_doan = None
    id_gan_nhat = -1

    for mid, x, nhan in D:
        if mid == skip_id:
            continue
        dist = kc_func(u, x)
        if dist < kc_min:
            kc_min = dist
            du_doan = nhan
            id_gan_nhat = mid
    return du_doan, id_gan_nhat

def main():
    try:
        with open("hoa30.txt", "r") as f:
            lines = [line.strip() for line in f if line.strip()]
    except FileNotFoundError:
        print("Khong the mo tep hoa30.txt", file=sys.stderr)
        return

    n, d = map(int, lines[0].split())
    D = []
    for i in range(1, n + 1):
        parts = lines[i].split()
        features = [float(x) for x in parts[:d]]
        nhan = parts[d]
        D.append((i, features, nhan))

    # 1. Phan loai mau (6.5, 3.0, 5.5, 2.0)
    u = [6.5, 3.0, 5.5, 2.0]
    e_nhan, e_id = phan_loai(u, D, kc_euclid)
    m_nhan, m_id = phan_loai(u, D, kc_manhattan)
    print("Mau (6.5, 3.0, 5.5, 2.0):")
    print(f"Euclid   : {e_nhan} (mau {e_id})")
    print(f"Manhattan: {m_nhan} (mau {m_id})\n")

    # 2. Danh gia bo mot mau
    dung_e = 0
    dung_m = 0
    for mid, x, nhan in D:
        pe_nhan, _ = phan_loai(x, D, kc_euclid, skip_id=mid)
        if pe_nhan == nhan:
            dung_e += 1

        pm_nhan, pm_id = phan_loai(x, D, kc_manhattan, skip_id=mid)
        if pm_nhan == nhan:
            dung_m += 1
        else:
            print(f"Manhattan du doan sai mau {mid} ({nhan} -> {pm_nhan}, gan mau {pm_id})")

    print(f"Do chinh xac Euclid   : {dung_e}/{n} = {dung_e / n * 100:.2f}%")
    print(f"Do chinh xac Manhattan: {dung_m}/{n} = {dung_m / n * 100:.2f}%")

if __name__ == "__main__":
    main()

File: tuan2/PhanC/bai48.jl
using Printf

struct Mau
    id::Int
    x::Vector{Float64}
    nhan::String
end

kc_euclid(u::Vector{Float64}, v::Vector{Float64}) = sqrt(sum((u .- v) .^ 2))
kc_manhattan(u::Vector{Float64}, v::Vector{Float64}) = sum(abs.(u .- v))

function phan_loai(u::Vector{Float64}, D::Vector{Mau}, kc_func::Function; skip_id::Int = -1)
    kc_min = Inf
    du_doan = ""
    id_gan_nhat = -1

    for m in D
        if m.id == skip_id continue end
        dist = kc_func(u, m.x)
        if dist < kc_min
            kc_min = dist
            du_doan = m.nhan
            id_gan_nhat = m.id
        end
    end
    return du_doan, id_gan_nhat
end

function main()
    if !isfile("hoa30.txt")
        println(stderr, "Khong the mo tep hoa30.txt")
        return
    end

    lines = filter(!isempty, strip.(readlines("hoa30.txt")))
    n, d = parse.(Int, split(lines[1]))

    D = Vector{Mau}(undef, n)
    for i in 1:n
        parts = split(lines[i + 1])
        x = parse.(Float64, parts[1:d])
        nhan = String(parts[d + 1])
        D[i] = Mau(i, x, nhan)
    end

    # 1. Phan loai mau (6.5, 3.0, 5.5, 2.0)
    u = [6.5, 3.0, 5.5, 2.0]
    e_nhan, e_id = phan_loai(u, D, kc_euclid)
    m_nhan, m_id = phan_loai(u, D, kc_manhattan)
    println("Mau (6.5, 3.0, 5.5, 2.0):")
    println("Euclid   : $e_nhan (mau $e_id)")
    println("Manhattan: $m_nhan (mau $m_id)\n")

    # 2. Danh gia bo mot mau
    dung_e = 0
    dung_m = 0
    for m in D
        pe_nhan, _ = phan_loai(m.x, D, kc_euclid, skip_id = m.id)
        if pe_nhan == m.nhan dung_e += 1 end

        pm_nhan, pm_id = phan_loai(m.x, D, kc_manhattan, skip_id = m.id)
        if pm_nhan == m.nhan
            dung_m += 1
        else
            println("Manhattan du doan sai mau $(m.id) ($(m.nhan) -> $pm_nhan, gan mau $pm_id)")
        end
    end

    @printf("Do chinh xac Euclid   : %d/%d = %.2f%%\n", dung_e, n, dung_e / n * 100)
    @printf("Do chinh xac Manhattan: %d/%d = %.2f%%\n", dung_m, n, dung_m / n * 100)
end

main()